Initial testing framework

n-dusan opened the pull request:
* Unit tests for: 
  * `upgrade_python_package`, 
  * `install_with_constraints`, 
  * `get_constraints_file_path`,
  *  `run_module`
* Integration tests:
  * Successful upgrade from `v2.0.0` -> `v2.0.1`
  * Unsuccessful upgrade from `v2.0.1` -> `v2.0.2`.  
    * Revert back to `v2.0.1`

n-dusan committed:
refact: raise errors and remove a condition

* Reraise errors that prevent upgrade-python-package from working, like
  if a local wheel isn't found.

* Remove redundant if statement

n-dusan committed:
feat: include `tox`

* Will probably be replaced with a more straightforward fixture that
  creates a virtual environment. Right now is used for local testing purposes

n-dusan committed:
feat: add initial testing framework

* Setup fixtures. [TODO]: fixtures for python venv.

* Unit tests for installing v2.0.0 top_level dependency.

n-dusan committed:
fix: support pinned .whl versions

n-dusan committed:
refact: remove tox

n-dusan committed:
feat: create and cleanup venv

*For each session, create a fresh venv and install required packages. VEnv is cleaned up after session.

n-dusan committed:
feat: create 2.0.1 whl versions

n-dusan committed:
refact: rename `upgrade_from_local_wheel` test suite

* Unit tests upgrading a python package from 2.0.0 -> 2.0.1
* Specify which dependencies are expected to exist in venv. For example
  `test_upgrade_local_wheel_top_level_package_2_0_0_where_package_name_is_valid_expect_package_installed`
  expects that oll-* packages are on 2.0.1 versions

n-dusan committed:
feat: test `run_python_module`

n-dusan committed:
wip: integ. tests

n-dusan committed:
fix: enable pinned versions to be run from module

n-dusan committed:
feat: `install_local_package` util, venv fixtures

n-dusan committed:
refact: update test readability

* By using sets, it's more clear what are expected dependencies

n-dusan committed:
feat: set module run tests output

* Use capsys to figure out whether module can be run or not.

n-dusan committed:
feat: add `constraints.txt` unit tests

* Unit tests that load requirements from constraints.txt for packages
  that are installed in a pytest virtual environment. These should match
  correctly, and there shouldn't exist any dependencies outside of the
  ones that are asserted.

feat: test `install_with_constraints`

* Installing correct v2.0.0 and v2.0.1 should work correctly. v2.0.2
  should fail, since there aren't any v2.0.2 wheels.

renatav commented on upgrade/tests/test_check_constraints_file_path.py:
```diff
@@ -0,0 +1,76 @@
+    lacks = set(expected) - set(actual)
+    extra = set(actual) - set(expected)
+    message = f"Lacks elements {lacks} " if lacks else ""
+    message += f"Extra elements {extra}" if extra else ""
```
Don't really need a nice error message if we are not doing anything with it. Just check len

renatav commented on upgrade/tests/test_check_constraints_file_path.py:
```diff
@@ -0,0 +1,76 @@
+
+    lacks = set(expected) - set(actual)
+    extra = set(actual) - set(expected)
+    message = f"Lacks elements {lacks} " if lacks else ""
```
Same

n-dusan committed:
refact: revert dependency version if pip fails

* If --no-deps passes, but the underlying dependencies aren't resolved
  by installing everything from `install_with_constraints`, we want to
  revert the dependency back to it's original state. This will only work
  if the dependency is installed prior to the script

n-dusan committed:
feat: mock constraints path

*Sometimes we need to not return a mocked constraints.txt file path, since no dependency is yet installed. This collides with a venv that's run on a local machine (in case an oll module exists on a local machine)

n-dusan committed:
feat: `v2.0.2` whl

*This is for top-level-package only. Installation should fail since oll-test-top-level depends on oll-dependency1 and oll-dependency2

n-dusan committed:
feat,refact: update upgrade py package tests

*
`test_upgrade_local_wheels_top_level_package_from_2_0_1_to_2_0_2_expect_error_and_revert_to_2_0_1`
shows that the upgrade py package scripts reverts to previous dependency
version.

n-dusan committed:
fix: consistency

renatav commented on upgrade/tests/test_check_constraints_file_path.py:
```diff
@@ -0,0 +1,64 @@
+
+
+def test_get_constraints_for_dependency1_from_venv_expect_constraints_exist_and_match():
+    full_package_name = "oll_dependency1-2.0.1-py2.py3-none-any.whl"
```
Can create a generic test here - wheel and expected as inputs since the 3 tests have the same code basically. Maybe with `parameterize`

n-dusan replied:
Makes sense

renatav commented on upgrade/tests/test_integ_upgrade_from_local_wheel.py:
```diff
@@ -52,3 +52,47 @@ def test_upgrade_local_wheel_top_level_package(wheels_dir, use_pip, capfd):
+    
+    assert f"Failed to install wheel {package}" in out
+    assert "Successfully uninstalled oll-test-top-level-2.0.2" in out
+    assert "Successfully installed oll-test-top-level-2.0.1" in out
```
awesome, looks good

n-dusan committed:
refact: parametrize check constraints file tests

n-dusan committed:
fix: mock `find_spec` to prevent venv leak

* Without `find_spec` mocked, function won't find the installed module
  from the fresh venv, but from existing venv from which pytest was
  invoked. This lead to module not being run, even though if it's valid.
  Mock `find_spec` to result to True, and assert if the module was run
  from unit tests.
  These tests were passing in the past, because the local development
  venv had oll-test-top-level installed. When testing this on a local
  machine, please uninstall oll-test-top-level.

n-dusan committed:
Testing framework update (#3)

* feat: add gitgub actions

* fix: set python executable path based on OS

* fix: rework getting constraints.txt in order to support Linux distributions

site-packages structure are not the same on Linux and Windows, so
the implementation has to take os into consideration

* refact: generic except

* Depending on the platform, except subprocess.CalledProcessError isn't
  caught. Make except generic

* refact: redundant test

* fix: drop `shell=True`

`shell=True` wasn't working on Ubuntu, and pip list didn't output anything, so drop the flag

* fix: restore `find_spec` after teardown

* refactor: avoid bare excepts

Co-authored-by: Renata <rvaderna@openlawlib.org>

renatav reviewed (approved):
The tests are passing both locally on my machine and on CI, so I think that we can merge now. Will do some manual testing, but we can address the issues in a new PR if they come up
